# Introducing tf.estimator

**Learning Objectives**
  - Understand where the `tf.estimator` module sits in the hierarchy of Tensorflow APIs
  - Understand the workflow of creating a `tf.estimator` model
    1. Create Feature Columns
    2. Create Input Functions
    3. Create Estimator
    4. Train/Evaluate/Predict
  - Understand how to swap in/out different types of Estimators
  
## Introduction 
Tensorflow is a hierarchical framework. The further down the hierarchy you go, the more flexibility you have, but that more code you have to write. Generally one starts at the highest level of abstraction. Then if you need additional flexibility drop down one layer.

<img src='../assets/TFHierarchy.png' width='50%'>
<sup>(image: https://www.tensorflow.org/guide/premade_estimators)</sup>

In this notebook we will be operating at the highest level of Tensorflow abstraction, using the Estimator API to predict taxifare prices on the sampled dataset we created previously.

In [1]:
import tensorflow as tf
import pandas as pd
import shutil

print(tf.__version__)
#tf.enable_eager_execution()

1.15.0


## Load raw data 

First let's download the raw .csv data. These are the same files created in the `create_datasets.ipynb` notebook

In [2]:
!gsutil cp gs://cloud-training-demos/taxifare/small/*.csv .
!ls -l *.csv

Copying gs://cloud-training-demos/taxifare/small/taxi-test.csv...
Copying gs://cloud-training-demos/taxifare/small/taxi-train.csv...              
Copying gs://cloud-training-demos/taxifare/small/taxi-valid.csv...              
- [3 files][ 10.9 MiB/ 10.9 MiB]                                                
Operation completed over 3 objects/10.9 MiB.                                     
-rw-r--r-- 1 jupyter jupyter 1799474 Nov 21 07:20 taxi-test.csv
-rw-r--r-- 1 jupyter jupyter 7986353 Nov 21 07:20 taxi-train.csv
-rw-r--r-- 1 jupyter jupyter 1673742 Nov 21 07:20 taxi-valid.csv
-rw-r--r-- 1 jupyter jupyter     163 Nov 21 06:46 test.csv


Because the files are small we can load them into in-memory Pandas dataframes.

In [3]:
df_train = pd.read_csv(filepath_or_buffer = "./taxi-train.csv")
df_valid = pd.read_csv(filepath_or_buffer = "./taxi-valid.csv")
df_test = pd.read_csv(filepath_or_buffer = "./taxi-test.csv")

CSV_COLUMN_NAMES = list(df_train)
print(CSV_COLUMN_NAMES)

FEATURE_NAMES = CSV_COLUMN_NAMES[1:] # all but first column
LABEL_NAME = CSV_COLUMN_NAMES[0] # first column
df_train

['fare_amount', 'dayofweek', 'hourofday', 'pickuplon', 'pickuplat', 'dropofflon', 'dropofflat']


fare_amount  dayofweek  hourofday  pickuplon  pickuplat  dropofflon  \
0               3.8          1          0 -73.986872  40.739698  -73.985832   
1               4.0          7          0 -73.978310  40.748493  -73.983726   
2              10.5          4          0 -73.990474  40.719493  -73.989566   
3              10.5          1          0 -73.990456  40.724912  -73.941419   
4               5.3          5          0 -73.974802  40.754947  -73.984790   
...             ...        ...        ...        ...        ...         ...   
151609         12.5          6         23 -74.003630  40.741817  -73.971822   
151610         14.5          5         23 -74.006580  40.739732  -73.971655   
151611         29.0          6         23 -73.982498  40.731735  -73.912415   
151612         11.7          7         23 -73.984452  40.746737  -73.996516   
151613         14.9          6         23 -73.949622  40.776805  -74.002024   

        dropofflat  
0        40.744041  
1        40.743893  
2        40.747033  
3        40.719180  
4        40.749647  
...            ...  
151609   40.762817  
151610   40.762760  
151611   40.764919  
151612   40.763236  
151613   40.739729  

[151614 rows x 7 columns]

## Create feature columns

Feature columns make it easy to perform common type of feature engineering on your raw data. For example you can one-hot encode categorical data, create feature crosses, embeddings and more. We'll cover these later in the course, but if you want to a sneak peak browse the official TensorFlow [feature columns guide](https://www.tensorflow.org/guide/feature_columns).

In our case we won't do any feature engineering. However we still need to create a list of feature columns because the Estimator we will use requires one. To specify the numeric values should be passed on without modification we use `tf.feature_column.numeric_column()`

#### **Exercise 1**

Use a [python list comprehension](https://www.pythonforbeginners.com/basics/list-comprehensions-in-python) or a `for` loop to create the feature columns for all features in `FEATURE_NAMES`.

In [12]:
feature_columns = [
    tf.feature_column.indicator_column(tf.feature_column.categorical_column_with_identity("dayofweek", 8)),  # TODO: fix
    tf.feature_column.indicator_column(tf.feature_column.categorical_column_with_identity("hourofday", 24)),
#    tf.feature_column.numeric_column("dayofweek"),
#    tf.feature_column.numeric_column("hourofday"),
    tf.feature_column.numeric_column("pickuplon"),
    tf.feature_column.numeric_column("pickuplat"),
    tf.feature_column.numeric_column("dropofflon"),
    tf.feature_column.numeric_column("dropofflat"),
]

## Define input function

Now that your estimator knows what type of data to expect and how to intepret it, you need to actually pass the data to it! This is the job of the input function.

The input function returns a new batch of (features, label) tuples each time it is called by the Estimator.

- features: A python dictionary. Each key is a feature column name and its value is the tensor containing the data for that feature
- label: A Tensor containing the labels

So how do we get from our current Pandas dataframes to (features, label) tuples that return one batch at a time?

The `tf.data` module contains a collection of classes that allows you to easily load data, manipulate it, and pipe it into your model. https://www.tensorflow.org/guide/datasets_for_estimators

#### **Exercise 2**

The code cell below has a few TODOs for you to complete. 

The first TODO in the `train_input_fn` asks you to create a tf.dataset using the [tf.data.Dataset](https://www.tensorflow.org/api_docs/python/tf/data/Dataset) API for input pipelines. Complete the code so that the variable `dataset` creates a tf.data.Dataset element using the [tf.from_tensor_slices method](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#from_tensor_slices). The argument `tensors` should be a tuple of a dict of the features and the label taken from the Pandas dataframe. 

The second TODO in the `train_input_fn` asks you to add a shuffle, repeat and batch operation to the dataset object you created above. Have a look at [the usage of these methods in the tf.data.Datasets API](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#methods)

The next TODO is in the `eval_input_fn`. Here you are asked to create a dataset object for the validation data. It should look similar to the pipeline you created for the `train_input_fn`. Note that for the `eval_input_fn` we don't add a shuffle or repeat step as we'll just evaluation a given batch during each validation step.

The last TODO is in the `predict_input_fn` where you are asked to once again use the Tensorflow Dataset API to set up a dataset for the prediction stage using the same `from_tensor_slices` as before. Note, during `PREDICT` we don't have the label, only features. 

In [13]:

#d = tf.data.Dataset.from_tensor_slices((dict(df_train[FEATURE_NAMES]), df_train[LABEL_NAME]))
#next(iter(d))

In [14]:
def train_input_fn(df, batch_size = 128):
    #1. Convert dataframe into correct (features, label) format for Estimator API
    dataset = tf.data.Dataset.from_tensor_slices((dict(df[FEATURE_NAMES]), df[LABEL_NAME]))
    
    # Note:
    # If we returned now, the Dataset would iterate over the data once  
    # in a fixed order, and only produce a single element at a time.
    
    #2. Shuffle, repeat, and batch the examples.
    dataset = dataset.repeat().shuffle(5128).batch(batch_size)
    # dataset = dataset.shuffle(5128).repeat().batch(batch_size)

    return dataset

def eval_input_fn(df, batch_size = 128):
    #1. Convert dataframe into correct (features, label) format for Estimator API
    dataset = tf.data.Dataset.from_tensor_slices((dict(df[FEATURE_NAMES]), df[LABEL_NAME]))
    
    #2.Batch the examples.
    dataset = dataset.batch(batch_size = batch_size)
   
    return dataset

def predict_input_fn(df, batch_size = 128):
    #1. Convert dataframe into correct (features) format for Estimator API
    dataset = tf.data.Dataset.from_tensor_slices((dict(df[FEATURE_NAMES])))

    #2.Batch the examples.
    dataset = dataset.batch(batch_size = batch_size)
   
    return dataset

## Choose Estimator

Tensorflow has several premade estimators for you to choose from:

- LinearClassifier/Regressor
- BoostedTreesClassifier/Regressor
- DNNClassifier/Regressor
- DNNLinearCombinedClassifier/Regressor

If none of these meet your needs you can implement a custom estimator using `tf.Keras`. We'll cover that later in the course.

For now we will use the premade LinearRegressor. To instantiate an estimator simply pass it what feature columns to expect and specify an directory for it to output checkpoint files to.

#### **Exercise 3**

Comlete the code in the cell below to define a Linear Regression model using the TF Estimator API. Have a [look at the documentation](https://www.tensorflow.org/api_docs/python/tf/estimator/LinearRegressor#__init__) to see what variables you must pass to initialize a `LinearRegressor` instance. You'll want to add values for `feature_columns`, `model_dir` and `config`. When setting up `config`, have a look at the [documentation for tf.estimator.RunConfig](https://www.tensorflow.org/api_docs/python/tf/estimator/RunConfig) and be sure to set `tf.random_seed` to ensure reproducibility.

In [7]:
OUTDIR = "taxi_trained"

model = tf.estimator.LinearRegressor(
    feature_columns, OUTDIR
)

INFO:tensorflow:Using default config.
INFO:tensorflow:Using config: {'_num_worker_replicas': 1, '_evaluation_master': '', '_tf_random_seed': None, '_experimental_max_worker_delay_secs': None, '_experimental_distribute': None, '_is_chief': True, '_protocol': None, '_eval_distribute': None, '_keep_checkpoint_every_n_hours': 10000, '_service': None, '_keep_checkpoint_max': 5, '_task_type': 'worker', '_session_config': allow_soft_placement: true
graph_options {
  rewrite_options {
    meta_optimizer_iterations: ONE
  }
}
, '_num_ps_replicas': 0, '_train_distribute': None, '_master': '', '_save_summary_steps': 100, '_save_checkpoints_secs': 600, '_task_id': 0, '_model_dir': 'taxi_trained', '_device_fn': None, '_save_checkpoints_steps': None, '_global_id_in_cluster': 0, '_cluster_spec': <tensorflow.python.training.server_lib.ClusterSpec object at 0x7f90a75664e0>, '_session_creation_timeout_secs': 7200, '_log_step_count_steps': 100}


## Train

Simply invoke the estimator's `train()` function. Specify the `input_fn` which tells it how to load in data, and specify the number of steps to train for.

By default estimators check the output directory for checkpoint files before beginning training, so it can pickup where it last left off. To prevent this we'll delete the output directory before starting training each time.

In [8]:
%%time
tf.logging.set_verbosity(tf.logging.INFO) # so loss is printed during training
shutil.rmtree(path = OUTDIR, ignore_errors = True) # start fresh each time

model.train(
    input_fn = lambda: train_input_fn(df = df_train), 
    steps = 500)

Instructions for updating:
Use Variable.read_value. Variables in 2.X are initialized automatically both in eager and graph (inside tf.defun) contexts.
INFO:tensorflow:Calling model_fn.
Instructions for updating:
Please use `layer.add_weight` method instead.
Instructions for updating:
If using Keras pass *_constraint arguments to layers.
Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where
Instructions for updating:
Use `tf.cast` instead.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Create CheckpointSaverHook.
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Saving checkpoints for 0 into taxi_trained/model.ckpt.
INFO:tensorflow:loss = 26450.172, step = 0
INFO:tensorflow:global_step/sec: 120.462
INFO:tensorflow:loss = 10155.403, step = 100 (0.835 sec)
INFO:tensorflow:global_step/sec: 155.164
INFO:tensorflow:loss = 15846.178, step = 200 (0.641 sec)
INFO

## Evaluate

Estimators similarly have an `evaluate()` function. In this case we don't need to specify the number of steps to train because we didn't tell our input function to repeat the data. Once the input function reaches the end of the data evaluation will end. 

Loss is reported as MSE by default so we take the square root before printing.

#### **Exercise 4**

Complete the code in the cell below to run evaluation on the model you just trained. You'll use the `evaluate` method of the `LinearRegressor` model you created and trained above. Have a look at [the documentation of the evaluate method](https://www.tensorflow.org/api_docs/python/tf/estimator/LinearRegressor#evaluate) here to see what it expects. Note you'll need to pass the evaluation input function as a lambda function processing the Pandas dataframe `df_valid`.

In [9]:
def print_rmse(model, df):
    metrics = model.evaluate(
        lambda: eval_input_fn(df)
    )
    print("RMSE on dataset = {}".format(metrics["average_loss"]**.5))
print_rmse(model = model, df = df_valid)

INFO:tensorflow:Calling model_fn.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Starting evaluation at 2019-11-21T07:21:10Z
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Restoring parameters from taxi_trained/model.ckpt-500
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Finished evaluation at 2019-11-21-07:21:12
INFO:tensorflow:Saving dict for global step 500: average_loss = 85.760864, global_step = 500, label/mean = 11.229713, loss = 10959.48, prediction/mean = 11.997164
INFO:tensorflow:Saving 'checkpoint_path' summary for global step 500: taxi_trained/model.ckpt-500
RMSE on dataset = 9.260716184929354


RMSE of 9.43 is  worse than our rules based benchmark (RMSE of $7.70). However given that we haven't done any feature engineering or hyperparameter tuning, and we're training on a small dataset using a simple linear model, we shouldn't yet expect good performance. 

The goal at this point is to demonstrate the mechanics of the Estimator API. In subsequent notebooks we'll improve on the model.

## Predict

To run prediction on the test set `df_test` we use the `predict_input_fn` you created above, passsing the `df_test` dataframe for prediction. We'll use our model to make predicitons on the first 10 elements of the `df_test` dataframe.

In [10]:
predictions = model.predict(input_fn = lambda: predict_input_fn(df = df_test[:10]))
for items in predictions:
    print(items)

INFO:tensorflow:Calling model_fn.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Restoring parameters from taxi_trained/model.ckpt-500
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
{'predictions': array([11.831713], dtype=float32)}
{'predictions': array([11.994401], dtype=float32)}
{'predictions': array([12.087061], dtype=float32)}
{'predictions': array([11.832011], dtype=float32)}
{'predictions': array([11.832211], dtype=float32)}
{'predictions': array([12.112196], dtype=float32)}
{'predictions': array([12.111671], dtype=float32)}
{'predictions': array([11.989036], dtype=float32)}
{'predictions': array([11.829121], dtype=float32)}
{'predictions': array([12.109522], dtype=float32)}


Further evidence of the primitiveness of our model, it predicts almost the same amount for every trip!

## Change Estimator type

One of the payoffs for using the Estimator API is we can swap in a different model type with just a few lines of code. Let's try a DNN. Note how now we need to specify the number of neurons in each hidden layer. Have a look at [the documentation for the DNN Regressor](https://www.tensorflow.org/api_docs/python/tf/estimator/DNNRegressor) to see what other variables you can set. 

In [15]:
%%time
tf.logging.set_verbosity(tf.logging.INFO)
shutil.rmtree(path = OUTDIR, ignore_errors = True)

model = tf.estimator.DNNRegressor(
    hidden_units = [2,3], # specify neural architecture
    feature_columns = feature_columns, 
    model_dir = OUTDIR,
    config = tf.estimator.RunConfig(tf_random_seed = 1)
)
model.train(
    input_fn = lambda: train_input_fn(df = df_train), 
    steps = 500)
print_rmse(model = model, df = df_valid)

INFO:tensorflow:Using config: {'_num_worker_replicas': 1, '_evaluation_master': '', '_tf_random_seed': 1, '_experimental_max_worker_delay_secs': None, '_experimental_distribute': None, '_is_chief': True, '_protocol': None, '_eval_distribute': None, '_keep_checkpoint_every_n_hours': 10000, '_service': None, '_keep_checkpoint_max': 5, '_task_type': 'worker', '_session_config': allow_soft_placement: true
graph_options {
  rewrite_options {
    meta_optimizer_iterations: ONE
  }
}
, '_num_ps_replicas': 0, '_train_distribute': None, '_master': '', '_save_summary_steps': 100, '_save_checkpoints_secs': 600, '_task_id': 0, '_model_dir': 'taxi_trained', '_device_fn': None, '_save_checkpoints_steps': None, '_global_id_in_cluster': 0, '_cluster_spec': <tensorflow.python.training.server_lib.ClusterSpec object at 0x7f8fed383f28>, '_session_creation_timeout_secs': 7200, '_log_step_count_steps': 100}
INFO:tensorflow:Calling model_fn.
Instructions for updating:
The old _FeatureColumn APIs are being de

Our performance is only slightly better at 9.26, and still far worse than our rules based model.  This illustrates an important tenant of machine learning: A more complex model can't outrun bad data. 

Currently since we're not doing any feature engineering our input data has very little signal to learn from, so using a DNN doesn't help much.

## Results summary

We can summarize our results in a table here.

#### **Exercise 5** 

Insert the results you found for the `LinearRegressor` and `DNNRegressor` model performance here.
   
|Model | RMSE on validation set|
|------|-----------------|
|Rules Based Benchmark| 7.76|
|Linear Model| 9.25 |
|DNN Model| 9.24 |

## Challenge exercise

Create a neural network that is capable of finding the volume of a cylinder given the radius of its base (r) and its height (h). Assume that the radius and height of the cylinder are both in the range 0.5 to 2.0. Simulate the necessary training dataset.
<p>
Hint (highlight to see):
<p style='color:white'>
The input features will be r and h and the label will be $\pi r^2 h$
Create random values for r and h and compute V.
Your dataset will consist of r, h and V.
Then, use a DNN regressor.
Make sure to generate enough data.
</p>

Copyright 2019 Google Inc. Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0 Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License